c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -0,0 +1,370 @@
+        ref r => panic!("unexpected resolution {:?} for `core::cell::Cell`", r),
+    };
+
+    let cell_adt = tcx.adt_def(cell_struct);
+    let substs = tcx.mk_substs([GenericArg::from(ty)].into_iter());
+    tcx.mk_adt(cell_adt, substs)
+}
+
+/// Produce a `Ty` reflecting the rewrites indicated by the labels in `rw_lty`.
+fn mk_rewritten_ty<'tcx>(lcx: LabeledTyCtxt<'tcx, RewriteLabel>, rw_lty: RwLTy<'tcx>) -> Ty<'tcx> {
+    let tcx = *lcx;
+    lcx.rewrite_unlabeled(rw_lty, &mut |ty, args, label| {
+        let (own, qty) = match label.ty_desc {
+            Some(x) => x,
+            None => return ty,
+        };
+
+        assert_eq!(args.len(), 1);
+        let mut ty = args[0];
+
+        if own == Ownership::Cell {
+            ty = mk_cell(tcx, ty);
+        }
+
+        ty = match qty {
+            Quantity::Single => ty,
+            Quantity::Slice => tcx.mk_slice(ty),
+            // TODO: This should generate `OffsetPtr<T>` rather than `&[T]`, but `OffsetPtr` is NYI
+            Quantity::OffsetPtr => tcx.mk_slice(ty),
+        };
+
+        ty = match own {
+            Ownership::Raw => tcx.mk_imm_ptr(ty),
+            Ownership::RawMut => tcx.mk_mut_ptr(ty),
+            Ownership::Imm => tcx.mk_imm_ref(tcx.mk_region(ReErased), ty),
+            Ownership::Cell => tcx.mk_imm_ref(tcx.mk_region(ReErased), ty),
+            Ownership::Mut => tcx.mk_mut_ref(tcx.mk_region(ReErased), ty),
+            Ownership::Rc => todo!(),
+            Ownership::Box => todo!(),
+        };
+
+        ty
+    })
+}
+
+struct HirTyVisitor<'a, 'tcx> {
+    acx: &'a AnalysisCtxt<'a, 'tcx>,
+    rw_lcx: LabeledTyCtxt<'tcx, RewriteLabel>,
+    //mir: &'a Body<'tcx>,
+    //span_index: SpanIndex<Location>,
+    hir_rewrites: Vec<(Span, Rewrite)>,
+}
+
+impl<'a, 'tcx> HirTyVisitor<'a, 'tcx> {
+    fn handle_ty(&mut self, rw_lty: RwLTy<'tcx>, hir_ty: &hir::Ty<'tcx>) {
+        if rw_lty.label.ty_desc.is_none() && !rw_lty.label.descendant_has_rewrite {
+            // No rewrites here or in any descendant of this HIR node.
+            return;
+        }
+
+        let hir_args = match deconstruct_hir_ty(rw_lty.ty, hir_ty) {
+            Some(x) => x,
+            None => {
+                // `hir_ty` doesn't have the expected structure (for example, we expected a type
+                // like `*mut T`, but it's actually an alias `MyPtr`), so we can't rewrite inside
+                // it.  Instead, we discard it completely and pretty-print `rw_lty` (with rewrites
+                // applied).
+                let ty = mk_rewritten_ty(self.rw_lcx, rw_lty);
+                let printer = FmtPrinter::new(*self.rw_lcx, Namespace::TypeNS);
+                let s = ty.print(printer).unwrap().into_buffer();
+                self.hir_rewrites.push((hir_ty.span, Rewrite::PrintTy(s)));
+                return;
+            }
+        };
+
+        if let Some((own, qty)) = rw_lty.label.ty_desc {
+            assert_eq!(hir_args.len(), 1);
```

Comment:
will this assertion break when we want to introduce ADT handling?

Reply:
also, what about tuples?

Reply:
`rw_lty.label.ty_desc` is only `Some` for pointer and reference types.  The `type_desc` module in general is only about pointers - its purpose is to convert pointer permissions into a matching safe reference type.